File: 2015/day8/ruby/logic.rb
# frozen_string_literal: true

def count_memory_chars(input)
  no_quotes = input[1..-2]
  two_letter_match_count = no_quotes.scan(/(\\")|(\\\\)/).length
  four_letter_match_count = no_quotes.scan(/(\\x[0-9|a-f|A-F]{2})/).length
  input.length - (no_quotes.length - (two_letter_match_count + (3 * four_letter_match_count)))
end

def count_reencode_string_chars(input)
  no_quotes = input[1..-2]
  escaped_slash_count = no_quotes.scan(/(\\)/).length
  escaped_quotes_count = no_quotes.scan(/(")/).length
  (escaped_slash_count + escaped_quotes_count) + 4
end

File: 2015/day8/go/logic/logic.go
package logic

import (
	"regexp"
	"strings"
)

var memoryCharsRegex = regexp.MustCompile(`(\\")|(\\\\)`)
var fourLetterCharsRegex = regexp.MustCompile(`(\\x[0-9|a-f|A-F]{2})`)
var escapedSlashCharRegex = regexp.MustCompile(`(\\)`)
var escapedQuoteCharRegex = regexp.MustCompile(`(")`)

func CountMemoryChars(input string) int {
	noQuotes := strings.Trim(input, "\"")
	twoLetterMatchCount := len(memoryCharsRegex.FindAllIndex([]byte(noQuotes), -1))
	fourLetterMatchCount := len(fourLetterCharsRegex.FindAllIndex([]byte(noQuotes), -1))
	return len(input) -
		(len(noQuotes) - (twoLetterMatchCount + (3 * fourLetterMatchCount)))
}

func CountReencodeStringChars(input string) int {
	escapedSlashCount := len(escapedSlashCharRegex.FindAllIndex([]byte(input), -1))
	escapedQuoteCount := len(escapedQuoteCharRegex.FindAllIndex([]byte(input), -1)) - 2
	return (escapedSlashCount + escapedQuoteCount) + 4
}
